components/SingleChat.js

```diff
@@ -39,124 +65,386 @@ class SingleChat extends Component {
+      recording.setOnRecordingStatusUpdate(this.updateRecordingStatus);
+      // set recording in constructor
+      this.recording = recording;
+      // begin recording
+      await this.recording.startAsync();
+
+      this.setState({
+        isLoading: false
+      });
+    } catch (err) {
+      console.log("Error starting recording: ", err);
+    }
+  }
+
+  // on stop recording click
+  async stopRecording() {
+    try {
+      this.setState({
+        isLoading: true
+      });
+      // stops recording and deallocates recorder from memory
+      await this.recording.stopAndUnloadAsync();
+      // customizes audio experience on iOS and Android
+      await setAudioMode({ allowsRecordingIOS: false });
+
+      if (this.recording) {
+        const audioUrl = this.recording.getURI();
+        this.recording.setOnRecordingStatusUpdate(null);
+        this.setState({ audioUrl });
+      }
+      // creates and loads a new sound object to play back the recording
+      const { sound, status } = await this.recording.createNewLoadedSoundAsync(
+        { isLooping: false },
+        this.updateSoundStatus
+      );
+      this.sound = sound;
+      this.setState({
+        isLoading: false
+      });
+    } catch (err) {
+      console.log("Error stopping recording: ", err);
+    }
+  }
+
+  // start / stop recording
+  handleRecordPressed() {
+    this.state.isRecording ? this.stopRecording() : this.startRecording();
+  }
+
+  // play / pause playback for recording
+  handleToggleRecording() {
+    if (this.recording != null) {
+      if (this.state.isRecordingPlaying) {
+        this.sound.pauseAsync();
+        this.setState({ isRecordingPlaying: false });
+      } else {
+        this.setState({ isRecordingPlaying: true });
+        this.sound.playAsync();
+      }
+    }
+  }
+
+  // dispatch send audio
+  async handleSendAudio() {
+    const { audioUrl } = this.state;
+    const fileName = `${genUUID()}.aac`;
+    const file = {
+      name: fileName,
+      type: "audio/aac",
+      uri: Platform.OS === "ios" ? audioUrl : `file://${audioUrl}`
+    };
+    const text = formatText(this.props);
+    text.messageType = "audio";
+    this.props.postAudio(file, text);
+  }
+
+  // dispatch send message
+  handleSendMessage(messages) {
+    let text = formatText(this.props);
+    text.message = messages[messages.length - 1].text;
+    text.messageType = "message";
+    this.props.postMessage(text);
+  }
+
   render() {
-    console.log(this.state);
     return (
       <View style={styles.container}>
         <Text style={{ color: "red" }}>{this.props.sendMessageError}</Text>
         <GiftedChat
           messages={this.props.messages}
           user={{
             _id: this.props.uid,
-            name: this.props.displayName,
+            name: this.props.displayName
           }}
-          onSend={this.handleSendMessage}
+          onSend={this.handleSendMessage.bind(this)}
```

Comment:
does "bind" need to be called here if it's already being called in the constructor? same with line 412